base/setup/lib/utils/filesup.c

```diff
@@ -71,78 +68,84 @@ SetupCreateSingleDirectory(
                           NULL,
                           0);
     if (NT_SUCCESS(Status))
-    {
         NtClose(DirectoryHandle);
-    }
-
-    RtlFreeUnicodeString(&PathName);
 
     return Status;
 }
 
 NTSTATUS
 SetupCreateDirectory(
-    IN PCWSTR PathName)
+    _In_ PCWSTR DirectoryName)
```

Comment:
You do create PathNameU from DirectoryName. Then you do search for separators in that. When looking the code below it seems to me as if DirectoryName would be allowed to contain separators. If this assumption is true, then I do find the word "Path" in the old parameters name to be more self-descriptionary than "Directory". Because within a "directory" I would not expect any spearators as a caller.
I do recommend to avoid the name "Directory" here. Maybe something like PassedPath?

Reply:
Yes indeed; I'll revert to "Path" naming. This is correct that "Path" is the whole thing with separators.

Reply:
>Isn't all access to the copy read-only as well? Why can't we use the passed parameter directly for all operations?

I made a copy because the string is modified at some point (when the path separators are being NULL'ed and restored); if one were to pass a static string residing in read-only memory this would lead to a crash (writing to read-only mem).

Reply:
Sorry my fault. I searched just for PathBuffer- and PanthNameU-names being referenced when searching for any writes. But I forgot that you could write to them also via the helper pointers. Ok, that explains the copy ofc.
Renaming the param is enough then to satisfy this conversation. Thx.

Reply:
I think that with @learn-more 's suggestions (having UNICODE_STRING versions of some of the functions being used) I _could_ avoid the need to allocate a string copy, because, instead of NULL-truncating at each path separator, I could just modify the UNICODE_STRING length (to count up to the separator).

Reply:
You could keep the non-unicode version, construct a UNICODE_STRING in that from the argument and call the Unicode version. (Note that it already resolves to `DoesPathExist`, which does construct a UNICODE_STRING).

Reply:
Yep, this is what I've done and this works well! (latest commit)